src/sys/futex.rs

```diff
@@ -0,0 +1,363 @@
+use crate::{Errno, Result};
+use libc::{syscall, SYS_futex};
+use std::cell::UnsafeCell;
+use std::convert::TryFrom;
+use std::os::unix::io::{FromRawFd, OwnedFd};
+use std::time::Duration;
+
+fn timespec(duration: Duration) -> libc::timespec {
+    let tv_sec = duration.as_secs().try_into().unwrap();
+    let tv_nsec = duration.subsec_nanos().try_into().unwrap();
+    libc::timespec { tv_sec, tv_nsec }
+}
+
+fn unwrap_or_null<T>(option: Option<&T>) -> *const T {
+    match option {
+        Some(t) => t,
+        None => std::ptr::null(),
+    }
+}
+
+/// Fast user-space locking.
+///
+/// By default we presume the futex is not process-private, that is, it is used across processes. If
+/// you know it is process-private you can set `PRIVATE` to `true`  which allows some additional
+/// optimizations.
+/// ```
+/// # use nix::{
+/// #   sys::{futex::Futex, mman::{mmap, MapFlags, ProtFlags}},
+/// #   errno::Errno,
+/// #   unistd::{fork,ForkResult},
+/// # };
+/// # use std::{
+/// #   time::{Instant, Duration},
+/// #   num::NonZeroUsize,
+/// #   mem::{ManuallyDrop, size_of},
+/// #   os::unix::io::OwnedFd,
+/// #   sync::Arc,
+/// #   thread::{spawn, sleep},
+/// # };
+/// const TIMEOUT: Duration = Duration::from_millis(500);
+/// const DELTA: Duration = Duration::from_millis(100);
+/// # fn main() -> nix::Result<()> {
+/// let futex: Futex = Futex::new(0);
+///
+/// // If the value of the futex is 0, wait for wake. Since the value is 0 and no wake occurs,
+/// // we expect the timeout will pass.
+///
+/// let instant = Instant::now();
+/// assert_eq!(futex.wait(0, Some(TIMEOUT)),Err(Errno::ETIMEDOUT));
```

Comment:
My familiarity is mostly with the PI variants of futexes, but my understanding is that proper futex use in general involves trying the operation in userspace via atomic operations before falling back to a syscall.

We should make sure the examples/tests demonstrate this, if only to avoid being incredibly confusing about their proper use.

Reply:
My understanding is that a `futex` is used like a Rust [`Condvar`](https://doc.rust-lang.org/stable/std/sync/struct.Condvar.html). I'm not sure what you mean by:

> proper futex use in general involves trying the operation in userspace via atomic operations before falling back to a syscall.

or how this would look.

Could you maybe provide an example?

Reply:
See https://linux.die.net/man/4/futex for the description (note it's futex(4), not the typical futex(2)).

The point of futexes is that syscalls are slow, but we can make the operations faster in uncontended cases by never leaving userspace. This is accomplished by doing a CAS operation and only performing the syscall if it fails.

See https://github.com/rtzoeller/rust-priority-inheriting-lock/blob/6225e90971d9aa72f070145788669e421cf012d4/src/lib.rs#L48 for an example of the userspace operation and falling back to the syscall for LOCK_PI (incidentally using linux-futex).

Reply:
I don't think this matter in non-highly contentious cases, not all use cases of a futex are highly contentious and not all are using it to implement a locking mechanism.

To have an example that properly illustrates what you suggest would be seem to need to be including an implementation of a priority-inheriting lock as an example.

Relating to docs, if you have suggestions for additions here please add a suggestion. Perhaps a comment suggesting that if a user is looking for a locking mechanism and does not require handling of priority-inheritance, they should use a mutex?

I'm not sure otherwise what changes you would suggest?